modules/controllers/cow-amm-controller.ts

```diff
@@ -33,104 +26,77 @@ export function CowAmmController(tracer?: any) {
+                toBlock = await subgraphClient.getMetadata().then((metadata) => metadata.block.number);
+            } catch (e) {
+                useSubgraph = false;
+            }
 
-                const allChangedPools = new Set<string>();
+            // Check if subgraph is not lagging behind
+            if (useSubgraph && Math.abs(latestBlock - toBlock) > acceptableSGLag) {
+                console.error(`CowAMM subgraph on ${chain} is lagging behind: ${latestBlock - toBlock}`);
+                useSubgraph = false;
+            }
 
-                for (let i = 0; i < numBatches; i++) {
-                    const from = fromBlock + (i > 0 ? 1 : 0) + i * rpcMaxBlockRange;
-                    const to = Math.min(fromBlock + (i + 1) * rpcMaxBlockRange, Number(toBlock));
-                    const changedPools = await fetchChangedPools(viemClient, chain, from, to);
-                    changedPools.forEach((pool) => allChangedPools.add(pool));
-                }
+            const ids: string[] = [];
 
-                if (allChangedPools.size === 0) {
-                    return [];
-                }
-                poolsToSync = Array.from(allChangedPools);
-            } else {
-                poolsToSync = await prisma.prismaPool
-                    .findMany({
-                        where: {
-                            chain,
-                            type: 'COW_AMM',
-                        },
-                        select: {
-                            id: true,
+            // Reload all pools if we are starting from the beginning
+            if (fromBlock === 0) {
+                const pools = await subgraphClient
+                    .getAllPools({ isInitialized: true })
+                    .then((pools) => pools.map((pool) => pool.id));
+                ids.push(...pools);
+            } else if (useSubgraph) {
+                const pools = await subgraphClient
+                    .getAllPools({
+                        isInitialized: true,
+                        _change_block: {
+                            number_gte: lastSyncBlock,
                         },
                     })
                     .then((pools) => pools.map((pool) => pool.id));
+                ids.push(...pools);
+            } else {
+                const pools = await fetchChangedPools(viemClient, chain, fromBlock, latestBlock, rpcMaxBlockRange);
+                ids.push(...pools);
             }
 
-            await upsertPools(poolsToSync, viemClient, subgraphClient, chain, toBlock);
-            await updateVolumeAndFees(chain, poolsToSync);
+            await upsertPools(ids, viemClient, subgraphClient, chain, latestBlock);
+            await updateVolumeAndFees(chain, ids);
             await updateSurplusAPRs();
+            // Update only SG blocks, so we don't miss any events
+            await upsertLastSyncedBlock(chain, PrismaLastBlockSyncedCategory.COW_AMM_POOLS, toBlock);
```

Comment:
Shouldnt this be updated after the balances sync?

Reply:
tried to make balances independent, but they aren't, so yes, moved that to the end